x/nexus/keeper/lockable_coin.go

```diff
@@ -0,0 +1,207 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibctypes "github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+
+	axelarnet "github.com/axelarnetwork/axelar-core/x/axelarnet/exported"
+	"github.com/axelarnetwork/axelar-core/x/nexus/exported"
+	"github.com/axelarnetwork/axelar-core/x/nexus/types"
+	"github.com/axelarnetwork/utils/funcs"
+)
+
+// NewLockableCoin creates a new lockable coin
+func (k Keeper) NewLockableCoin(ctx sdk.Context, ibc types.IBCKeeper, bank types.BankKeeper, coin sdk.Coin) (exported.LockableCoin, error) {
+	return newLockableCoin(ctx, k, ibc, bank, coin)
+}
+
+// lockableCoin provides functionality to lock and release coins
+type lockableCoin struct {
+	sdk.Coin
+	coinType types.CoinType
+	nexus    types.Nexus
+	ibc      types.IBCKeeper
+	bank     types.BankKeeper
+}
+
+// newLockableCoin creates a coin struct, assign a coin type and normalize the denom if it's a ICS20 token
+func newLockableCoin(ctx sdk.Context, nexus types.Nexus, ibc types.IBCKeeper, bank types.BankKeeper, coin sdk.Coin) (lockableCoin, error) {
+	coinType, err := getCoinType(ctx, nexus, coin.GetDenom())
+	if err != nil {
+		return lockableCoin{}, err
+	}
+
+	// If coin type is ICS20, we need to normalize it to convert from 'ibc/{hash}'
+	// to native asset denom so that nexus could recognize it
+	if coinType == types.ICS20 {
+		denomTrace, err := ibc.ParseIBCDenom(ctx, coin.GetDenom())
+		if err != nil {
+			return lockableCoin{}, err
+		}
+
+		coin = sdk.NewCoin(denomTrace.GetBaseDenom(), coin.Amount)
+	}
+
+	c := lockableCoin{
+		Coin:     coin,
+		coinType: coinType,
+		nexus:    nexus,
+		ibc:      ibc,
+		bank:     bank,
+	}
+	if _, err := c.getOriginalCoin(ctx); err != nil {
```

Comment:
we should add a check that this derives the same asset denom provided originally. Currently, IBC denom parsing ignore the qualified path, and thus could be a spoofed denom